### DISCUSSION TITLE
執行錯誤：TypeError: Cannot read property 'id' of undefined

### DISCUSSION THREAD

**Discussion by weih210287** [Lidemy/forum#93] [其他問題] [answered]
各位大大好，想請問一個問題，為甚麼在迴圈裡使用 `json[i].id`, `json[i].name` 會發生這種錯誤呢？
在寫 hw1 要印出排行前 10 名的書時，我使用了迴圈來印出每本書的編號與名稱，然而在迴圈外印出 `json[i].id` 都沒問題，可是只要一丟進迴圈，雖然結果跑得出來，但最後都還是會出現 Cannot read property of undefined 的錯誤，如圖：

![](https://i.imgur.com/uO1lale.jpg)

而且就算解決了編號的問題，跑到 `json[i].name` 時還是會有一樣的錯誤，上網爬了文說好像是因為程式找不到我要的物件 ID，可是結果明明就跑得出來所以又讓我更疑惑了 QWQ
以下是我的程式碼：
```javascript
const request = require('request')

request(
  'https://lidemy-book-store.herokuapp.com/books?_limit=10',
  function(err, response, body) {
    const json = JSON.parse(body)
    for (let i = 0; i <= 10; i++) {
      console.log(`${json[i].id} ${json[i].name}`)
    }
  }
)
```
美好的早晨就從卡關開始 TvT 謝謝大家的幫忙

**Comment by allenliaooooooo** [comment, marked as answer]
你迴圈的條件沒寫好哦~

**Reply by weih210287** [reply]
喔喔感謝感謝！
一直想著要印出 10 本就粗心了
